File: examples/java/src/main/java/examples/copilot/ExplainCodeFile.java
// 04 — Explain a code file
//
// Read a file from disk and ask Copilot for a plain-English explanation.
//
// Run: mvn -q exec:java -Dexec.mainClass=examples.copilot.ExplainCodeFile \
//        -Dexec.args="src/main/java/examples/copilot/HelloWorld.java"
package examples.copilot;

import com.github.copilot.sdk.CopilotClient;
import com.github.copilot.sdk.CopilotSession;
import com.github.copilot.sdk.generated.AssistantMessageEvent;
import com.github.copilot.sdk.json.CopilotClientOptions;
import com.github.copilot.sdk.json.MessageOptions;
import com.github.copilot.sdk.json.PermissionHandler;
import com.github.copilot.sdk.json.SessionConfig;

import java.nio.file.Files;
import java.nio.file.Path;

public final class ExplainCodeFile {
  private static String resolveCopilotCli() {
      String pathEnv = System.getenv("PATH");
      if (pathEnv != null) {
          for (String dir : pathEnv.split(java.io.File.pathSeparator)) {
              java.io.File candidate = new java.io.File(dir, "copilot");
              if (candidate.canExecute()) {
                  return candidate.getAbsolutePath();
              }
          }
      }
      String envOverride = System.getenv("COPILOT_CLI_PATH");
      if (envOverride != null && !envOverride.isEmpty()) {
          return envOverride;
      }
      throw new IllegalStateException(
          "copilot CLI not found. Install @github/copilot globally or set COPILOT_CLI_PATH."
      );
  }

  public static void main(String[] args) throws Exception {
    if (args.length < 1 || args[0].isBlank()) {
      System.err.println("usage: ExplainCodeFile <file>");
      System.exit(1);
    }

    Path target = Path.of(args[0]);
    String source = Files.readString(target);

    String prompt = String.join("\n",
        "Explain what this file does in 4-6 bullet points.",
        "Call out anything subtle (race conditions, hidden side effects, error swallowing).",
        "",
        "--- " + target + " ---",
        source
    );

    try (CopilotClient client = new CopilotClient(
        new CopilotClientOptions()
            .setCliPath(resolveCopilotCli())
            .setCliArgs(new String[]{"--disable-builtin-mcps"})
    )) {
      client.start().get();

      CopilotSession session = client.createSession(
          new SessionConfig()
              .setModel("gpt-4.1")
              .setOnPermissionRequest(PermissionHandler.APPROVE_ALL)
      ).get();

      AssistantMessageEvent response = session.sendAndWait(
          new MessageOptions().setPrompt(prompt)
      ).get();

      System.out.println(response.getData().content());
    }
  }
}

File: examples/rust/src/bin/explain_code_file.rs
//! 04 — Explain a code file
//!
//! Read a file from disk and ask Copilot for a plain-English explanation.
//!
//! Run: cargo run --bin explain_code_file -- <path-to-source-file>
//! e.g. cargo run --bin explain_code_file -- src/bin/hello_world.rs

use github_copilot_sdk::{CliProgram, Client, ClientOptions, SessionConfig};
use std::env;
use std::fs;
use std::path::PathBuf;

fn resolve_copilot_cli() -> std::path::PathBuf {
    // PATH lookup first
    if let Ok(path_var) = std::env::var("PATH") {
        for dir in path_var.split(':') {
            let candidate = std::path::Path::new(dir).join("copilot");
            if candidate.is_file() {
                return candidate;
            }
        }
    }
    // Fall back to COPILOT_CLI_PATH env var
    if let Ok(env_path) = std::env::var("COPILOT_CLI_PATH") {
        if !env_path.is_empty() {
            return std::path::PathBuf::from(env_path);
        }
    }
    panic!("copilot CLI not found. Install @github/copilot globally or set COPILOT_CLI_PATH.");
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args: Vec<String> = env::args().collect();
    if args.len() < 2 || args[1].is_empty() {
        eprintln!("usage: cargo run --bin explain_code_file -- <file>");
        std::process::exit(1);
    }
    let target = PathBuf::from(&args[1]);
    let source = fs::read_to_string(&target)?;

    let prompt = [
        "Explain what this file does in 4-6 bullet points.".to_string(),
        "Call out anything subtle (race conditions, hidden side effects, error swallowing)."
            .to_string(),
        String::new(),
        format!("--- {} ---", target.display()),
        source,
    ]
    .join("\n");

    let opts = ClientOptions::default()
        .with_program(CliProgram::Path(resolve_copilot_cli()))
        .with_extra_args(["--disable-builtin-mcps"]);
    let client = Client::start(opts).await?;
    let mut config = SessionConfig::default();
    config.model = Some("gpt-4.1".to_string());
    let config = config.approve_all_permissions();
    let session = client.create_session(config).await?;

    if let Some(event) = session.send_and_wait(prompt).await? {
        if let Some(content) = event.data.get("content").and_then(|v| v.as_str()) {
            println!("{}", content);
        }
    }

    session.destroy().await?;
    client.stop().await?;
    Ok(())
}
